driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -65,6 +67,30 @@ public static TimeoutContext createMaintenanceTimeoutContext(final TimeoutSettin
         return new TimeoutContext(true, timeoutSettings, calculateTimeout(timeoutSettings.getTimeoutMS()));
     }
 
+    public static TimeoutContext createTimeoutContext(final ClientSession session, final TimeoutSettings timeoutSettings) {
+        TimeoutContext sessionTimeoutContext = session.getTimeoutContext();
+
+        if (sessionTimeoutContext != null) {
+            TimeoutSettings sessionTimeoutSettings = sessionTimeoutContext.timeoutSettings;
+            if (timeoutSettings.getGenerationId() > sessionTimeoutSettings.getGenerationId()) {
+                throw new MongoClientException("Cannot change the timeoutMS during a transaction.");
+            }
+
+            // Check for any legacy operation timeouts
+            if (sessionTimeoutSettings.getTimeoutMS() == null) {
```

Comment:
`maxAwaitTimeMS`, which can be set on an `Iterable`, isn't currently taken into account in this logic when user is using "withTransaction" or explicit client sessions, remaining at 0 regardless of user settings.